oktadeveloper/okta-aws-cli-assume-role
# :bug: Correctly handle nested configs

Problem Statement
-----------------

Issue #141 states:
> Existing `~/.aws/config` with nested properties are overwritten improperly by this tool.
> For example, `s3` configuration uses [indentation for its properties](https://docs.aws.amazon.com/cli/latest/topic/s3-config.html) e.g.
> ```
> [profile development]
> aws_access_key_id=foo
> aws_secret_access_key=bar
> s3 =
>   max_concurrent_requests = 20
>   max_queue_size = 10000
> ```
> When `awscli` is run, the `config` is written out like so:
> ```
> [profile development]
> aws_access_key_id=foo
> aws_secret_access_key=bar
> s3 =
> max_concurrent_requests = 20
> max_queue_size = 10000
> ```
> which is not correctly read by `aws s3` commands!
> 
> Specifically, the following test https://github.com/oktadeveloper/okta-aws-cli-assume-role/blob/054c45c5b0b58195db45bd25b528c0436331d7ed/src/test/java/com/okta/tools/aws/settings/ConfigurationTest.java#L110-L130 if modified to contain s3 keys as follows:
> ```java
>     @Test
>     void addOrUpdateProfileToExistingProfile() throws IOException {
>         final String existingCombined = manualRole + "\n"
>                 + "s3 =\n"
>                 + "    max_queue_size = 1000" + "\n\n" + existingProfile;
> 
>         final StringReader configurationReader = new StringReader(existingCombined);
>         final StringWriter configurationWriter = new StringWriter();
>         final Configuration configuration = new Configuration(configurationReader);
> 
>         final String updatedPrefix = "updated_";
>         final String expected = "[profile " + profileName + "]\n"
>                 + Configuration.ROLE_ARN + " = " + updatedPrefix + role_arn + "\n"
>                 + SOURCE_PROFILE + " = " + profileName + "_source\n"
>                 + Configuration.REGION + " = " + region + "\n"
>                 + "s3 =\n"
>                 + "    max_queue_size = 1000"
>                 + "\n\n" + existingProfile;
> 
> 
>         configuration.addOrUpdateProfile(profileName, updatedPrefix + role_arn);
>         configuration.save(configurationWriter);
> 
>         String given = StringUtils.remove(configurationWriter.toString().trim(), '\r');
> 
>         assertEquals(expected, given);
>     }
> ```
> will not pass.

Solution
--------
 - Add Apache Commons Configuration2

 - Adapt configuration2's INIConfiguration to support AWS quirks

 - Remove Ini4j (unmaintained, has park page for website)

 - Refactor file handling code

Resolves #141

Apologies for the extensiveness of this change. The configuration
handling likely needs a refactor to make it easier to support future
modification.
